pkg/apis/v1alpha1/cluster_deployment_template.go

```diff
@@ -41,6 +66,41 @@ type ClusterDeploymentTemplateList struct {
 	Items           []ClusterDeploymentTemplate `json:"items"`
 }
 
+var _ webhook.Validator = &ClusterSourceTemplate{}
+
+func (c *ClusterDeploymentTemplate) ValidateCreate() error {
+	return c.validate()
+}
+
+func (c *ClusterDeploymentTemplate) ValidateUpdate(_ runtime.Object) error {
+	return c.validate()
+}
+
+func (c *ClusterDeploymentTemplate) ValidateDelete() error {
+	return nil
+}
+
+func (c *ClusterDeploymentTemplate) validate() error {
+	err := c.Spec.TemplateSpec.validate()
+	if err != nil {
+		return err
+	}
+
+	if bothSet(c.Spec) || neitherSet(c.Spec) {
+		return fmt.Errorf("invalid spec: must set exactly one of spec.ObservedMatches and spec.ObservedCompletion")
+	}
+
+	return nil
+}
+
+func bothSet(spec DeploymentSpec) bool {
```

Comment:
`bothSet` and `neitherSet` here end up being available (although privately) for the whole v1alpha1 package. given that _it is_ DeploymentSpec specific (given the signature of the functions), I wonder if it'd be better to instead make them DeploymentSpec methods instead 

```suggestion
func (spec DeploymentSpec) both_somethingsomething_Set() bool {
```

wdyt?

Reply:
Good call